backend/de.metas.contracts/src/main/java/de/metas/contracts/flatrate/process/C_Flatrate_Conditions_Extend.java

```diff
@@ -0,0 +1,94 @@
+import org.compiere.model.I_C_Year;
+import org.compiere.util.Ini;
+
+import static de.metas.contracts.model.X_C_Flatrate_Conditions.TYPE_CONDITIONS_ModularContract;
+import static de.metas.contracts.model.X_C_Flatrate_Conditions.ONFLATRATETERMEXTEND_ExtensionNotAllowed;
+
+public class C_Flatrate_Conditions_Extend extends JavaProcess implements IProcessPrecondition
+{
+	public final static String MSG_FLATRATE_CONDITIONS_EXTENSION_NOT_ALLOWED = "@MSG_FLATRATE_CONDITIONS_EXTENSION_NOT_ALLOWED@";
+	public final static String MSG_SETTINGS_WITH_SAME_YEAR_ALREADY_EXISTS = "@MSG_SETTINGS_WITH_SAME_YEAR_ALREADY_EXISTS@";
+	private final IFlatrateDAO flatrateDAO = Services.get(IFlatrateDAO.class);
+	private final IFlatrateBL flatrateBL = Services.get(IFlatrateBL.class);
+
+	@Param(parameterName = I_ModCntr_Settings.COLUMNNAME_C_Year_ID, mandatory = true)
+	private int p_C_Year_ID;
+
+	@Override
+	protected String doIt()
+	{
+		final I_C_Year newYear = InterfaceWrapperHelper.load(YearId.ofRepoId(p_C_Year_ID), I_C_Year.class);
+		final I_C_Flatrate_Conditions conditions = flatrateDAO.getConditionsById(ConditionsId.ofRepoId(getRecord_ID()));
+
+		if (isSettingsWithSameYear(conditions.getModCntr_Settings(), newYear))
+		{
+			throw new AdempiereException(MSG_SETTINGS_WITH_SAME_YEAR_ALREADY_EXISTS);
+		}
+
+		final I_C_Flatrate_Conditions newConditions = flatrateBL.extendConditionsToNewYear(conditions, newYear);
+		if (newConditions != null)
+		{
+			final int adWindowId = getProcessInfo().getAD_Window_ID();
+			final TableRecordReference ref = TableRecordReference.of(newConditions);
+			if (adWindowId > 0 && !Ini.isSwingClient())
+			{
+				getResult().setRecordToOpen(ref, adWindowId, OpenTarget.SingleDocument);
+			}
+		}
+
+		return MSG_OK;
+	}
+
+	private boolean isSettingsWithSameYear(@NonNull final I_ModCntr_Settings settings, @NonNull final I_C_Year newYear)
+	{
+		return settings.getC_Year_ID() == newYear.getC_Year_ID();
+	}
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(IProcessPreconditionsContext context)
+	{
+		if (!context.isSingleSelection())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNotSingleSelection();
+		}
+
+		if (!I_C_Flatrate_Conditions.Table_Name.equals(context.getTableName()))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("The process only runs with C_Flatrate_Conditions table");
+		}
+
+		final I_C_Flatrate_Conditions currentConditions = context.getSelectedModel(I_C_Flatrate_Conditions.class);
+
+		if (!TYPE_CONDITIONS_ModularContract.equals(currentConditions.getType_Conditions()))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("The process runs only  when contract of type  ModularContract");
+		}
+
+		if (!ONFLATRATETERMEXTEND_ExtensionNotAllowed.equals(currentConditions.getOnFlatrateTermExtend()))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason(MSG_FLATRATE_CONDITIONS_EXTENSION_NOT_ALLOWED);
+		}
+
```

Comment:
In fact, this should also be done via process parameter, so that the C_Year_ID only shows years relevant to the currently selected calendar (via ModCntr_Settings.C_Calendar__ID)